### PR TITLE
December work

### PR DESCRIPTION
**Pull request by petebryan**
New features:
- Added driver for MDATP advanced hunting queries with associated queries
- Created parent OData driver for MDATP and Security Graph
- Added data connector for querying data from Azure on subscriptions and resources
Fixes:
- Other minor fixes

### CODE REVIEW

File: msticpy/data/queries/kql_mdatp_user.yaml [hunk 1]
```diff
@@ -0,0 +1,80 @@
...
+    args:
+      query: '
+        LogonEvents
+        | where EventTime >= datetime({start})
+        | where EventTime <= datetime({end})
+        | where AccountName has "{username}"
+        {add_query_items}'
+      uri: None
+    parameters:
+      username:
```
Comment (ianhelle) [inline msticpy/data/queries/kql_mdatp_user.yaml]: account_name?

File: msticpy/data/queries/kql_mdatp_user.yaml [hunk 2]
```diff
@@ -0,0 +1,80 @@
...
+    metadata:
+    args:
+      query: '
+        ProcessCreationEvents
+        | where EventTime >= datetime({start})
+        | where EventTime <= datetime({end})
+        | where AccountName has "{username}"
+        {add_query_items}'
+    parameters:
+      username:
```
Comment (ianhelle) [inline msticpy/data/queries/kql_mdatp_user.yaml]: account_name?

File: msticpy/data/queries/kql_mdatp_user.yaml [hunk 3]
```diff
@@ -0,0 +1,80 @@
...
+    metadata:
+    args:
+      query: '
+        FileCreationEvents
+        | where EventTime >= datetime({start})
+        | where EventTime <= datetime({end})
+        | where InitiatingProcessAccountName contains "{username}"
+        {add_query_items}'
+    parameters:
+      username:
```
Comment (ianhelle) [inline msticpy/data/queries/kql_mdatp_user.yaml]: account_name?

File: msticpy/data/queries/kql_mdatp_user.yaml [hunk 4]
```diff
@@ -0,0 +1,80 @@
...
+    metadata:
+    args:
+      query: '
+        NetworkCommunicationEvents
+        | where EventTime >= datetime({start})
+        | where EventTime <= datetime({end})
+        | where InitiatingProcessAccountName contains "{username}"
+        {add_query_items}'
+    parameters:
+      username:
```
Comment (ianhelle) [inline msticpy/data/queries/kql_mdatp_user.yaml]: account_name?

File: msticpy/data/queries/kql_mdatp_user.yaml [hunk 5]
```diff
@@ -0,0 +1,80 @@
...
+        default: -30
+      end:
+        description: Query end time
+        type: datetime
+        default: 0
+      add_query_items:
+        description: Additional query clauses
+        type: str
+        default: ''
+sources:
```
Comment (ianhelle) [inline msticpy/data/queries/kql_mdatp_user.yaml]: Was think that you could make some of these more flexible by including default params for host_name, account_name, file_path and then doing some Kusto logic to exclude things that are not set by the user. I have an example in kql_sent_alert.list_related_alerts - it's a bit klunky but might save some typing in the future.

File: msticpy/data/azure_data.py [hunk 6]
```diff
@@ -0,0 +1,299 @@
...
+            resource = self.resource_client.resources.get(
+                resource_details["resource_group_name"],
+                resource_details["resource_provider_namespace"],
+                resource_details["parent_resource_path"],
+                resource_details["resource_type"],
+                resource_details["resource_name"],
+                self.get_api(resource_id),
+            )
+
+        resource_details = {
```
[the remark this answers is not in the document]
Comment (petebryan) [reply msticpy/data/azure_data.py]: I like this approach - implemented.

File: msticpy/data/azure_data.py [hunk 7]
```diff
@@ -0,0 +1,299 @@
...
+        self.resource_client = ResourceManagementClient(self.credentials, sub_id)
+        if rgroup is None:
+            resources = self.resource_client.resources.list()
+        else:
+            resources = self.resource_client.resources.list_by_resource_group(rgroup)
+
+        if get_props is True:
+            print("Collecting properties for every resource may take some time...")
+
+        ids = []
```
[the remark this answers is not in the document]
Comment (petebryan) [reply msticpy/data/azure_data.py]: Went with the attr class suggestion

File: msticpy/data/queries/kql_mdatp_alerts.yaml [hunk 8]
```diff
@@ -0,0 +1,94 @@
...
+    metadata:
+    args:
+      query: '
+        {table}
+        | where EventTime >= datetime({start})
+        | where EventTime <= datetime({end})
+        | where ComputerName has "{hostname}"
+        {add_query_items}'
+    parameters:
+      hostname:
```
[the remark this answers is not in the document]
Comment (petebryan) [reply msticpy/data/queries/kql_mdatp_alerts.yaml]: Changes

File: msticpy/data/queries/kql_mdatp_alerts.yaml [hunk 9]
```diff
@@ -0,0 +1,94 @@
...
+    metadata:
+    args:
+      query: '
+        {table}
+        | where EventTime >= datetime({start})
+        | where EventTime <= datetime({end})
+        | where SHA1 has "{sha1}"
+        {add_query_items}'
+    parameters:
+      sha1:
```
[the remark this answers is not in the document]
Comment (petebryan) [reply msticpy/data/queries/kql_mdatp_alerts.yaml]: Yes the AlertEvents table schema has only SHA1

File: msticpy/data/queries/kql_mdatp_network.yaml [hunk 10]
```diff
@@ -0,0 +1,94 @@
...
+        | where EventTime >= datetime({start})
+        | where EventTime <= datetime({end})
+        | where RemoteUrl has "{url}"
+        {add_query_items}'
+    parameters:
+      url:
+        description: Remote URL
+        type: str
+  protocol_connections:
+    description: Lists alerts associated with a specified SHA1 hash
```
[the remark this answers is not in the document]
Comment (petebryan) [reply msticpy/data/queries/kql_mdatp_network.yaml]: Fixed

File: msticpy/data/queries/kql_mdatp_process.yaml [hunk 11]
```diff
@@ -0,0 +1,84 @@
...
+    args:
+      query: '
+        {table}
+        | where EventTime >= datetime({start})
+        | where EventTime <= datetime({end})
+        | where ComputerName has "{hostname}"
+        {add_query_items}'
+      uri: None
+    parameters:
+      hostname:
```
[the remark this answers is not in the document]
Comment (petebryan) [reply msticpy/data/queries/kql_mdatp_process.yaml]: Changed

File: msticpy/data/queries/kql_mdatp_process.yaml [hunk 12]
```diff
@@ -0,0 +1,84 @@
...
+        type: str
+  process_creations:
+    description: Lists all processes created by name or hash
+    metadata:
+    args:
+      query: '
+        {table}
+        | where EventTime >= datetime({start})
+        | where EventTime <= datetime({end})
+        | where FileName contains "{processid}" or SHA1 has "{processid}" or SHA256 has "{processid}" or MD5 has "{processid}"
```
[the remark this answers is not in the document]
Comment (petebryan) [reply msticpy/data/queries/kql_mdatp_process.yaml]: renamed to process_identifier

File: msticpy/data/queries/kql_mdatp_process.yaml [hunk 13]
```diff
@@ -0,0 +1,84 @@
...
+    metadata:
+    args:
+      query: '
+        {table}
+        | where EventTime >= datetime({start})
+        | where EventTime <= datetime({end})
+        | where FileName contains "{path}"
+        {add_query_items}'
+    parameters:
+      path:
```
[the remark this answers is not in the document]
Comment (petebryan) [reply msticpy/data/queries/kql_mdatp_process.yaml]: Changed

File: msticpy/data/queries/kql_mdatp_user.yaml (hunk 1 above)
Comment (petebryan) [reply msticpy/data/queries/kql_mdatp_user.yaml]: changed

File: msticpy/data/queries/kql_mdatp_user.yaml (hunk 2 above)
Comment (petebryan) [reply msticpy/data/queries/kql_mdatp_user.yaml]: changed

File: msticpy/data/queries/kql_mdatp_user.yaml (hunk 4 above)
Comment (petebryan) [reply msticpy/data/queries/kql_mdatp_user.yaml]: changed

File: msticpy/data/queries/kql_mdatp_user.yaml (hunk 3 above)
Comment (petebryan) [reply msticpy/data/queries/kql_mdatp_user.yaml]: changed

File: msticpy/data/azure_data.py [hunk 14]
```diff
@@ -0,0 +1,300 @@
...
+            The subscription ID to get resources for
+        rgroup: str (Optional)
+            The name of a Resource Group to get resources for
+        get_props: bool (Optional)
+            Set to True if you want to get the full properties of every resource
+            Warning this may be a slow process depending on the number of resources
+
+        Returns
+        -------
+        resrouce_df: pd.DataFrame
```
Comment (ashwin-patil) [inline msticpy/data/azure_data.py]: small typo in docstrings..  resource_df

File: msticpy/data/azure_data.py [hunk 15]
```diff
@@ -0,0 +1,300 @@
...
+
+# pylint: disable=too-few-public-methods
+# attr class doesn't need a method
+@attr.s
+class Items:
+    """attr class to build resource details dictionary."""
+
+    resouce_id = attr.ib()
+    name = attr.ib()
+    ressource_type = attr.ib()
```
Comment (ashwin-patil) [inline msticpy/data/azure_data.py]: typo ? should it be resource_type ?

**Pull request merged**